silkworm/db/datastore/snapshots/snapshot_repository.cpp

```diff
@@ -128,37 +140,46 @@ void SnapshotRepository::reopen_folder() {
     auto file_ranges = list_dir_file_ranges();
     if (file_ranges.empty()) return;
 
-    // sort file_ranges by range.start first, then by range size
-    std::ranges::sort(file_ranges, [](const StepRange& r1, const StepRange& r2) -> bool {
-        if (r1.start != r2.start) {
-            return r1.start < r2.start;
-        }
-        return r1.size() > r2.size();
-    });
+    ThreadPool worker_pool;
 
     std::unique_lock lock(*bundles_mutex_);
-    // copy bundles prior to modification
-    auto bundles = std::make_shared<Bundles>(*bundles_);
 
-    Step num = file_ranges[0].start;
+    std::vector<std::future<std::shared_ptr<SnapshotBundle>>> completed_future_bundles;
     for (const auto& range : file_ranges) {
-        // avoid gaps/overlaps
-        if (range.start != num) continue;
         if (range.size() == 0) continue;
 
-        if (!bundles->contains(num)) {
-            SnapshotBundlePaths bundle_paths{schema_, dir_path_, range};
-            // if all bundle paths exist
-            if (std::ranges::all_of(bundle_paths.files(), [](const fs::path& p) { return fs::exists(p); })) {
-                SnapshotBundle bundle = open_bundle(range);
-                bundles->insert_or_assign(num, std::make_shared<SnapshotBundle>(std::move(bundle)));
-                // avoid gaps/overlaps
-                num = range.end;
+        SnapshotBundlePaths bundle_paths{schema_, dir_path_, range};
+        // Open iff all bundle paths exist
+        if (std::ranges::all_of(bundle_paths.files(), [](const fs::path& p) { return fs::exists(p); })) {
+            // Schedule each bundle opening on worker pool collecting its future result for completion handling
+            completed_future_bundles.emplace_back(worker_pool.submit([&, range]() -> std::shared_ptr<SnapshotBundle> {
+                return std::make_shared<SnapshotBundle>(schema_, dir_path_, range, index_salt_);
+            }));
+        }
+    }
+    while (!completed_future_bundles.empty()) {  // all bundles will be eventually opened
+        std::erase_if(completed_future_bundles, [&](auto& future_bundle_ptr) {
+            if (future_bundle_ptr.wait_for(std::chrono::seconds(0)) == std::future_status::ready) {
+                std::shared_ptr<SnapshotBundle> bundle_ptr = future_bundle_ptr.get();
+                const auto step_range = bundle_ptr->step_range();
+                bundles_->insert_or_assign(step_range.start, std::move(bundle_ptr));
+                return true;
             }
+            return false;
+        });
+    }
+
+#ifndef NDEBUG
+    // Sanity check: no gap must exist in bundles
+    if (!bundles_->empty()) {
+        Step max_end = bundles_->cbegin()->second->step_range().start;
+        for (const auto& b : *bundles_) {
+            SILKWORM_ASSERT(b.second->step_range().start == max_end);
+            max_end = b.second->step_range().end;
         }
     }
```

Comment:
This is optional, but std::adjacent_find seems like a perfect algorithm here. It takes a predicate that accepts 2 arguments: entry and entry_next. Note that `entry.first` is always equal to `entry.second->step_range().start` in this map. The predicate can compare entry.second->step_range().end to entry_next.first. 

With C++ 23 [pairwise view](https://en.cppreference.com/w/cpp/ranges/adjacent_view) this code would be even simpler.

Reply:
I'll leave it for later